torch/tensor.py

```diff
@@ -722,4 +722,25 @@ def _update_names(self, names, inplace):
         else:
             return super(Tensor, self).rename(names)
 
+    @property
+    def grad(self):
+        """
+        This attribute is ``None`` by default and becomes a Tensor the first time a call to
+        :func:`backward` computes gradients for ``self``.
+        The attribute will then contain the gradients computed and future calls to
+        :func:`backward` will accumulate (add) gradients into it.
+        """
+        if self.requires_grad and not hasattr(self, "retains_grad") and not self.is_leaf and self._grad is None:
+            warnings.warn("Accessing .grad on a Tensor that is not a leaf Tensor: its .grad attribute won't "
+                          "be populated during autograd.backward(). See .retain_grad() if you want the gradient to be saved.")
```

Comment:
I've googled "pytorch retain_grad" and didn't find more details about his. Maybe you could simply add the details to the pull request, and link to the pull request instead?

Without this, I suspect a user may not know how to fix this if there's something to fix after reading just "See .retain_grad() if you want the gradient to be saved."

Thoughts?

Reply:
The [doc](https://pytorch.org/docs/stable/tensors.html#torch.Tensor.retain_grad) is indeed pretty limited.
Should I add an example of what it does to the doc and state explicitly to check the pytorch doc in the warning?

Reply:
I see you could add the information to anyone of the following, with an explicit link in the warning.

* #30531
* [doc](https://pytorch.org/docs/stable/tensors.html#torch.Tensor.retain_grad)
* updated doc in #30253

The first seems natural since it links to other related issues and gives you space to write details. Thoughts?

Reply:
I updated the comment to link this PR.